src/index.ts

```diff
@@ -1,176 +1,102 @@
- * @description Queries the Manager Topic for all TLD messages and stores them
- */
-  populateTLDMessages = async () => {
-    this.tldMessages = await queryTLDTopicMessages();
-  };
+  public async init() {
+    this.topLevelDomains = await this.getTopLevelDomains();
+  }
 
   /**
- * @description Generate a NameHash of the provided domain string
- * @param domain: {string} The domain string to hash
- * @returns {Buffer}
+ * @description Resolves a Second Level Domain to the wallet address of the domain's owner
+ * @param domain: {string} The domain to query
+ * @returns {Promise<AccountId>}
  */
-  static generateNameHash = (domain: string): NameHash => {
-    if (!domain) {
-      return {
-        domain,
-        tldHash: Buffer.from([0x0]),
-        sldHash: Buffer.from([0x0]),
-      };
-    }
-    const domainsList = domain.split('.').reverse();
-    const tld = domainsList[0];
-    let sld;
-    if (domainsList.length > 1) {
-      sld = domainsList.slice(0, 2);
+  public async resolveSLD(domain: string): Promise<string> {
+    try {
+      const nameHash = hashDomain(domain);
+      const sld = await this.getSecondLevelDomain(nameHash);
+      const [tokenId, serial] = sld.nftId.split(':');
+      const nft = await this.mirrorNode.getNFT(tokenId, serial);
+      return nft.account_id;
+    } catch (err) {
+      throw err;
     }
+  };
 
-    let tldHash = Buffer.from([0x0]);
-    let sldHash = Buffer.from([0x0]);
-
-    if (tld) {
-      tldHash = keccak256(tld);
-    }
-    if (sld) {
-      sldHash = sld.reduce(
-        (prev, curr) => keccak256(prev + curr),
-        Buffer.from(''),
-      );
-    }
+  // Private
 
-    return { domain, tldHash, sldHash };
-  };
+  /**
+ * @description Retrieves and stores top level domains
+ */
+  private async getTopLevelDomains(): Promise<TopLevelDomain[]> {
+    const response = await this.mirrorNode.getTopicMessages(tldTopicId, 0);
+    const messages: MessageObject[] = response;
+
+    return messages.map((messageObject: MessageObject) => {
+      const decoded = Buffer.from(messageObject.message, 'base64').toString();
+      return JSON.parse(decoded) as TopLevelDomain;
+    });
+  }
 
   /**
    * @description Get the tld message on the Manager topic for a given nameHash
    * @param nameHash: {NameHash} The nameHash for the sld to query
    * @returns {Promise<TLDTopicMessage>}
    */
-  private queryTLDTopicMessage = async (nameHash: NameHash): Promise<TLDTopicMessage> => {
-    try {
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      const found = this.tldMessages.find((message: any) => (message.nameHash.tldHash === nameHash.tldHash.toString('hex')));
-      if (!found) throw new Error('Not Found');
+  private getTopLevelDomain(nameHash: NameHash): TopLevelDomain {
+    const found = this.topLevelDomains.find((tld: TopLevelDomain) => (
+      tld.nameHash.tldHash === nameHash.tldHash.toString('hex')
+    ));
+    if (!found) throw new Error('TLD not found');
```

Comment:
for the sdk we should have errors that extend off of js error. This allows users of the sdk to catch specific error to have better error handling